### PR TITLE
ASP.NET config builders

### LINKED ISSUE #8184: Document ASP.NET config builders
We should move the content in https://github.com/aspnet/MicrosoftConfigurationBuilders/blob/master/README.md into the official ASP.NET docs.

### PR DESCRIPTION
**Pull request by Rick-Anderson**
Fixes #8184
[Internal review URL](https://review.docs.microsoft.com/en-us/aspnet/config-builder?branch=pr-en-us-8374)

### CODE REVIEW

**Comment by Rick-Anderson** [comment]
@StephenMolloy please review. [Internal review URL](https://review.docs.microsoft.com/en-us/aspnet/config-builder?branch=pr-en-us-8374)

**Comment by scottaddie** [comment]
@Rick-Anderson In this case, the first commit is to blame: https://github.com/aspnet/Docs/pull/8374/commits/ef7feb90bec55085b8b59054067c0036175ad999. Notice it includes "WIP" in the commit message.

**Comment by Rick-Anderson** [comment]
@StephenMolloy 	can you give this another review? 

> it might be worth having a short section about best practices wrt keeping your config data secure. 

Can you send me a draft of that and where you suggest I add it? I'd like to get this published this week.

File: aspnet/config-builder.md [hunk 1]
```diff
@@ -0,0 +1,309 @@
...
+ms.date: 10/9/2018
+ms.technology: aspnet
+msc.type: content
+---
+
+# Configuration builders for ASP.NET
+
+By [Stephen Molloy](https://github.com/StephenMolloy) and [Rick Anderson](https://twitter.com/RickAndMSFT)
+
+Configuration builders provide a mechanism for ASP.NET apps to get configuration values.
```
Comment (csharpfritz) [inline aspnet/config-builder.md]: can we add: "... to get configuration values _from external sources_"

File: aspnet/config-builder.md [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+            }
+        },
+
+        "connectionStrings" : {
+            "myJSON_ConnectionString" : "See security warning."
+        }
+    }
+```
+
+> [!WARNING]
```
Comment (csharpfritz) [inline aspnet/config-builder.md]: Should this warning section be with the UserSecretsConfigurationBuilder?

**Comment by csharpfritz** [comment]
Added a few comments, did not run the samples as they appear correct and I'm assuming someone else has already run them in a real application

**Comment by Rick-Anderson** [comment]
@blowdart 	I'm inclined to remove the connection string from the following sample JSON:

```
{
        "appSettings" : {
            "setting1" : "value1",
            "setting2" : "value2",
            "complex" : {
                "setting1" : "complex:value1",
                "setting2" : "complex:value2",
            }
        },

        "connectionStrings" : {
            "myJSON_ConnectionString" : "See security warning."
        }
    }
```

We can't show insecure code and then tell folks not to use it. Misquoting Eilon - Folks that need to do this will be smart enough to figure it out without a sample.

@csharpfritz  or @blowdart 	 or @StephenMolloy 	: Can you provide me a draft of the following?

> it might be worth having a short section about best practices wrt keeping your config data secure.

@csharpfritz appreciate the review, I took all your suggestions.

**Comment by StephenMolloy** [comment]
I don't have any official draft of anything. But generally speaking, config builders are about enabling more modern and agile configuration in old school .Net. They _can_ be used in ways that promote more secure patterns and practices... but config builders themselves are not a security feature. As with any feature, developers must assess their needs and balance them against potential threats when designing a secure solution. If there is some bit of information that is too sensitive to have right there in the config file, then it should not be kept in an equally insecure plain text file in the application - so don't use a config builder that does that.

Unfortunately, these Key/Value config builders only operate on two sections... one of which is connectionStrings which commonly contain sensitive information. I just think its worth calling out that config builders are not a security feature by themselves. They still have to be deployed smartly with security in mind. I think the warning at the start of the json and user sectrest config builder sections are pretty good indicators that these two builders aren't necessarily secure by default. For the json sample, it seems like a good idea to follow Eilon's misquoted advice and not include a connection string in that one. ;)

File: aspnet/config-builder.md [hunk 3]
```diff
@@ -13,13 +13,16 @@ msc.type: content
...
 
-Configuration builders provide a mechanism for ASP.NET apps to get configuration values from external sources.
+Configuration builders provide a modern and agile mechanism for ASP.NET apps to get configuration values from external sources.
 
 Configuration builders:
 
 * Are available in .Net Framework .Net 4.7.1 and later.
 * Provide a flexible mechanism for reading configuration values.
 * Address some of the basic needs of apps as they move into a container and cloud focused environment.
+* Operate on two sections, `<appSettings />` and `<connectionStrings />`.
```
Comment (StephenMolloy) [inline aspnet/config-builder.md]: "Configuration Builders" as a concept - the concept that is introduced in .Net 4.7.1 - apply to any config section. Not just appSettings and connectionStrings.

They set of key/value config builders that we ship via nuget are paritally constrained to those two sections. In token-replace mode (mode="Expand") can be used on any config section.

**Comment by Rick-Anderson** [comment]
@StephenMolloy 	are we ready to merge?

**Comment by StephenMolloy** [comment]
I just gave the overall doc a last look over instead of peeking through individual commits. A few more suggestions:

**Comment by StephenMolloy** [comment]
I just gave the overall doc a last look over instead of peeking through individual commits. The technical details look good and more comprehensive. Just a few nits around phrasing where we could be a little more clear about the difference between key/value builders in Microsoft.Configuration.ConfigurationBuilders and the concept of ConfigurationBuilders in .Net, as well as being careful to not position config builders as a security feature by itself. It was too weird making edits by vague human-readable reference in this comment, so I just edited the raw .md and will send it to you via email.

**Comment by Rick-Anderson** [comment]
@StephenMolloy 	Nice updates. See the last two commits (your edits and my edits of your edits). I think this is ready to publish.

**Pull request merged**